File: Java/BaekJoon/Gold/G2_4195.java
import java.io.*;
import java.util.*;

public class Main {
    static HashMap<String, String> friends;
    static HashMap<String, Integer> nums;

    public static String find(String f) {
        if (friends.get(f).equals(f)) return f;

        String root = find(friends.get(f));
        friends.put(f, root);

        return root;
    }

    public static void merge(String a, String b) {
        String f1 = find(a);
        String f2 = find(b);

        if (f1.equals(f2)) return;

        friends.put(f2, f1);

        nums.put(f1, nums.get(f1) + nums.get(f2));
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int test = Integer.parseInt(br.readLine());

        for (int t = 0; t < test; t++) {
            friends = new HashMap<>();
            nums = new HashMap<>();

            int F = Integer.parseInt(br.readLine());

            for (int f = 0; f < F; f++) {
                StringTokenizer st = new StringTokenizer(br.readLine());

                String temp1 = st.nextToken();
                String temp2 = st.nextToken();

                if (!friends.containsKey(temp1)) {
                    friends.put(temp1, temp1);
                    nums.put(temp1, 1);
                }

                if (!friends.containsKey(temp2)) {
                    friends.put(temp2, temp2);
                    nums.put(temp2, 1);
                }

                merge(temp1, temp2);

                int result = nums.get(find(temp1));
                System.out.println(result);
            }
        }
    }
}

File: Cpp/BaekJoon/Gold/G2_4195.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

unordered_map<string, string> friends;
unordered_map<string, int> friends_num;

string find(string f) {
  if (friends[f] == f) return f;

  return friends[f] = find(friends[f]);
}

void merge(string f1, string f2) {
  string temp1 = find(f1);
  string temp2 = find(f2);

  if (temp1 == temp2) return;

  friends[temp2] = temp1;
  friends_num[temp1] += friends_num[temp2];
}

int main() {
  int test;
  cin >> test;

  for (int t = 0; t < test; t++) {
    friends.clear();
    friends_num.clear();

    int F;
    cin >> F;

    for (int f = 0; f < F; f++) {
      string f1, f2;
      cin >> f1 >> f2;

      if (friends.count(f1) == 0) {
        friends.insert({f1, f1});
        friends_num.insert({f1, 1});
      }

      if (friends.count(f2) == 0) {
        friends.insert({f2, f2});
        friends_num.insert({f2, 1});
      }

      merge(f1, f2);

      cout << friends_num[friends[f1]] << "\n";
    }
  }

  return 0;
}